### PR TITLE
Initial skeleton with recursive filepath parsing.

### PR DESCRIPTION
**Pull request by evetion**

### CODE REVIEW

File: hydrolib/core/basemodel.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    @classmethod
+    def validate(cls: Type["FileModel"], value: Any) -> "FileModel":
+        # Enable initialization with a Path.
+        if isinstance(value, Path):
+            # Pydantic Model init requires a dict
+            value = {"filepath": value}
+        return super().validate(value)
+
+    def _load(self, filepath: Path):
+        if Path(filepath).is_file():
```
Comment (priscavdsluis) [inline hydrolib/core/basemodel.py:59]:
```suggestion
        if filepath.is_file():
```

Can't we do the above?

File: hydrolib/core/basemodel.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+    def save(self, folder: Path, force=False):
+        """Save model and child models to their set filepaths.
+
+        For models with an unset filepath, we generate one based
+        on the given `folder`.
+
+        If `force` is set, we override the folder part of
+        already set filepaths, which can be used to copy complete models.
+        """
```
Comment (BeardedPlatypus) [inline hydrolib/core/basemodel.py:73]:
```suggestion
        """Save model and child models to their set filepaths.

        For models with an unset filepath, we generate one based
        on the given `folder`.

        If `force` is set, we override the folder part of
        already set filepaths, which can be used to copy complete models.
        
        Args:
            folder (Path): path to the folder where this FileModel will be stored
            force (bool, optional): Force override the folder part of already set filepath
        """
```

Do we need to document the arguments? or are they too trivial and do we ignore them?

File: hydrolib/core/basemodel.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        self._get_serializer()(self.filepath, data)
+
+    @classmethod
+    def _parse(cls, path: Path):
+        return cls._get_parser()(path)
+
+    @classmethod
+    def _generate_name(cls):
+        name, ext = cls._filename(), cls._ext()
+        return Path(f"{name}{ext}")
```
Comment (BeardedPlatypus) [inline hydrolib/core/basemodel.py:111]: Should we specify this returns a Path, or should this return str?

File: hydrolib/core/basemodel.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+            self.filepath = folder / self.filepath.name
+
+        # Convert child FileModels first
+        exclude = {"filepath"}
+        filemodel_fields = {}
+        for name, value in self:
+            if isinstance(value, FileModel):
+                filepath = value.save(folder, force)
+                filemodel_fields[name] = filepath
+                exclude.add(name)
```
Comment (priscavdsluis) [inline hydrolib/core/basemodel.py:88]: Does this prevent that the reference to this child file will not be written to the parent file? For example, in the mdu file we reference the network file.

**Review by BeardedPlatypus — CHANGES_REQUESTED** [review]

File: tests/test_model.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
+from pathlib import Path
+
+from hydrolib.core.basemodel import FileModel
+from hydrolib.core.models import DIMR, FMModel
+
+from .utils import test_data_dir, test_output_dir
+
+
+def test_filemodel_load():
+    d = DIMR(filepath="a", model=Path(test_data_dir / "test.mdu"))
```
Comment (priscavdsluis) [inline tests/test_model.py:10]: Did you intend to test with a file that does not exist?

**Review by priscavdsluis — CHANGES_REQUESTED** [review]

File: hydrolib/core/basemodel.py (hunk 4 above)
Comment (evetion) [reply hydrolib/core/basemodel.py:88]: It prevents the child file being returned as {fielda: value, etc} (automatic due to Pydantic), now it is returned as Path("location/of/child") by ourselves.

File: tests/test_model.py (hunk 5 above)
Comment (evetion) [reply tests/test_model.py:10]: Yes, I will make this explicit in the test description.

**Review by BeardedPlatypus — APPROVED** [review]

**Review by priscavdsluis — APPROVED** [review]

**Pull request merged**